[Flake] pubsub unit test times out with strange error

https://source.cloud.google.com/results/invocations/228df956-41e5-4c58-8631-d4f95292854c

```
306/306 Test #370: pubsub_internal_batching_publisher_connection_test .................***Timeout 1500.02 sec
Running main() from gmock_main.cc
[==========] Running 10 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 10 tests from BatchingPublisherConnectionTest
[ RUN      ] BatchingPublisherConnectionTest.DefaultMakesProgress
[       OK ] BatchingPublisherConnectionTest.DefaultMakesProgress (58 ms)
[ RUN      ] BatchingPublisherConnectionTest.BatchByMessageCount
[       OK ] BatchingPublisherConnectionTest.BatchByMessageCount (18 ms)
[ RUN      ] BatchingPublisherConnectionTest.BatchByMessageSize
[       OK ] BatchingPublisherConnectionTest.BatchByMessageSize (23 ms)
[ RUN      ] BatchingPublisherConnectionTest.BatchByMaximumHoldTime
[       OK ] BatchingPublisherConnectionTest.BatchByMaximumHoldTime (12 ms)
[ RUN      ] BatchingPublisherConnectionTest.BatchByFlush
/Volumes/BuildData/tmpfs/src/github/google-cloud-cpp/google/cloud/pubsub/internal/batching_publisher_connection_test.cc:250: Failure
Expected equality of these values:
  2
  request.messages_size()
    Which is: 1
[libprotobuf FATAL /Volumes/BuildData/tmpfs/src/github/google-cloud-cpp/cmake-out/macos/external/include/google/protobuf/repeated_field.h:1537] CHECK failed: (index) < (current_size_): 
/Volumes/BuildData/tmpfs/src/github/google-cloud-cpp/google/cloud/pubsub/internal/batching_publisher_connection_test.cc:289: Failure
Expected equality of these values:
  "test-message-id-" + std::to_string(i)
    Which is: "test-message-id-1"
  *r
    Which is: "ack-for-test-data-1"


99% tests passed, 1 tests failed out of 306

Total Test time (real) = 1514.54 sec

The following tests FAILED:
	370 - pubsub_internal_batching_publisher_connection_test (Timeout)
Errors while running CTest
make[2]: *** [build/google-cloud-cpp/src/google_cloud_cpp_project-stamp/google_cloud_cpp_project-test] Error 8
make[1]: *** [CMakeFiles/google_cloud_cpp_project.dir/all] Error 2
make: *** [all] Error 2
2020-10-29T14:28:00Z: build script reported errors. 
```

--- a/google/cloud/pubsub/internal/batching_publisher_connection_test.cc
+++ b/google/cloud/pubsub/internal/batching_publisher_connection_test.cc
@@ -208,14 +208,15 @@ TEST(BatchingPublisherConnectionTest, BatchByMaximumHoldTime) {
         return make_ready_future(make_status_or(response));
       });
 
-  google::cloud::internal::AutomaticallyCreatedBackgroundThreads background;
-  auto const ordering_key = std::string{};
+  // Start with an inactive message queue, to avoid flakes due to scheduling
+  // problems.
+  CompletionQueue cq;
   auto publisher = BatchingPublisherConnection::Create(
       topic,
       pubsub::PublisherOptions{}
           .set_maximum_hold_time(std::chrono::milliseconds(5))
           .set_maximum_batch_message_count(4),
-      ordering_key, mock, background.cq(), pubsub_testing::TestRetryPolicy(),
+      /*ordering_key=*/{}, mock, cq, pubsub_testing::TestRetryPolicy(),
       pubsub_testing::TestBackoffPolicy());
   auto r0 =
       publisher
@@ -234,8 +235,15 @@ TEST(BatchingPublisherConnectionTest, BatchByMaximumHoldTime) {
             EXPECT_EQ("test-message-id-1", *r);
           });
 
+  // Now that the two messages are queued, we can activate the completion queue.
+  // It should flush the messages in about 5ms.
+  std::thread t{[](CompletionQueue cq) { cq.Run(); }, cq};
+
   r0.get();
   r1.get();
+
+  cq.Shutdown();
+  t.join();
 }
 
 TEST(BatchingPublisherConnectionTest, BatchByFlush) {

